content/blogposts/2022/ranking-in-a-week.md

```diff
@@ -0,0 +1,120 @@
+
+For the past three years at Sourcegraph, I have been amassing a giant compiler-accurate index of source code. And like a giant dork, that giant pile of data has just been sitting there enabling only the most basic of code navigation operations: go to definition and find references. The ranking effort is our first official foray into consuming this data in a more interesting way.
+
+**And we've acquired a taste for it.**
+
+Ranking is a only small amuse-bouche of what we have planned in the future.
+
+### High-level overview
+
+PageRank is an algorithm created by Google co-founder Larry Page in 1998 to assign a numeric value to each webpage indexed by the search engine. The algorithm takes as input a graph representing the set of webpages and links between them, and outputs a probability of a user landing on a particular webpage (for every webpage) by randomly clicking on links. Webpages with many or highly _relevant_ links will have a higher PageRank score, and are presented higher in the list of matching search results.
+
+We utilize the PageRank algorithm to rank source code results in a similar way. In our current implementation, we construct a graph of _source code text documents_ where an edge between two documents indicates a reference of a symbol (variable, function, type, etc) defined in another file.
+
+<Figure
+    src="https://storage.googleapis.com/sourcegraph-assets/blog/new-search-ranking/PageRank-hero-illustration.png"
+    alt="PageRank graph"
+    caption="An graph of text documents indicating references between files."
+    />
+
+PageRank values calculated over this graph "flow" to the text documents that have the highest concentration of _code use_ or _re-use_. This tends to surface results that have high relevancy to a search query by bringing popular definitions and representative usages to the top.
+
+Larry Page himself states that _"the computation of PageRank is fairly straightforward if we ignore the issues of scale"_, and the Wikipedia article for the algorithm is filled with descriptions of [eigenvector centrality](https://en.wikipedia.org/wiki/Eigenvector_centrality) of a [stochastic matrix](https://en.wikipedia.org/wiki/Stochastic_matrix) via [Von Mises iteration](https://en.wikipedia.org/wiki/Power_iteration). To get ranking results as quickly as possible so we can validate and iterate on the properties of the graph structure itself, we made the pragmatic choice to use [Apache Spark](https://spark.apache.org/), as it provides [several implementations of PageRank](https://spark.apache.org/docs/1.6.1/api/java/org/apache/spark/graphx/lib/PageRank.html) out-of-the-box (please see [xkcd #353](https://xkcd.com/353/)).
```

Comment:
love the xkcd reference, very accurately captures the decision to use Spark

Reply:
we are monkeys hitting computers with bones